File: scaffold/treatments.py
"""Treatment loader for benchmark experiments.

Treatments are loaded from YAML configuration and built into Treatment objects
with fully-resolved skill configurations.

Usage:
    from scaffold.treatments import load_treatments, load_treatment

    treatments = load_treatments()  # Load all from tests/treatments.yaml
    treatment = load_treatment("SEPARATE_NAMES")
"""

import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import yaml

from skills.parser import load_skill_variant, skill_config

TREATMENTS_FILE = Path(__file__).parent.parent / "tests" / "treatments.yaml"
SKILL_BASE = Path(__file__).parent.parent / "skills" / "benchmarks"


@dataclass
class TreatmentConfig:
    """Configuration for a treatment loaded from YAML."""

    name: str
    description: str
    claude_md: str = ""
    skills: list[dict[str, Any]] = field(default_factory=list)


def _add_language_suffix(content: str, lang: str) -> str:
    """Add language suffix to frontmatter description."""
    suffix = "(Python)" if lang == "py" else "(TypeScript)"
    return re.sub(
        r'^(description: "?)(.+?)("?)$',
        rf"\1\2 {suffix}\3",
        content,
        count=1,
        flags=re.MULTILINE,
    )


def _filter_related_skills(sections: list[str]) -> list[str]:
    """Filter out related_skills sections."""
    return [s for s in sections if s and "<related_skills>" not in s]


def _build_skill_config(
    skill_dir: str,
    variant: str = "all",
    suffix: bool = False,
    include_related: bool = False,
    noise: bool = False,
) -> dict:
    """Build a skill configuration from a skill directory.

    Args:
        skill_dir: Skill directory name (e.g., "langsmith_trace")
        variant: Language variant to load (py, ts, all)
        suffix: Whether to add language suffix to description
        include_related: Whether to include <related_skills> sections
        noise: Whether this is a noise/distractor skill (simpler loading)

    Returns:
        Skill configuration dict for scaffold
    """
    skill_path = SKILL_BASE / skill_dir

    if noise:
        # Noise skills are simple - just load skill.md
        skill_md = skill_path / "skill.md"
        if skill_md.exists():
            return skill_config([skill_md.read_text()], None, None)
        return None

    # Load skill variant
    skill = load_skill_variant(skill_path, variant)
    sections = skill["all"]

    # Filter related_skills sections unless explicitly included
    if not include_related:
        sections = _filter_related_skills(sections)

    content = "\n\n".join(sections)

    # Add language suffix to description if requested
    if suffix and variant in ("py", "ts"):
        content = _add_language_suffix(content, variant)

    return skill_config([content], skill["scripts_dir"], skill["script_filter"])


def build_treatment_skills(skill_configs: list[dict[str, Any]]) -> dict[str, dict]:
    """Build skill dict from YAML skill configurations.

    Args:
        skill_configs: List of skill config dicts from YAML

    Returns:
        Dict mapping skill names to skill configurations
    """
    skills = {}

    for cfg in skill_configs:
        skill_dir = cfg.get("skill")
        name = cfg.get("name", skill_dir.replace("_", "-"))
        variant = cfg.get("variant", "all")
        suffix = cfg.get("suffix", False)
        include_related = cfg.get("include_related", False)
        noise = cfg.get("noise", False)

        skill_config = _build_skill_config(
            skill_dir=skill_dir,
            variant=variant,
            suffix=suffix,
            include_related=include_related,
            noise=noise,
        )

        if skill_config:
            skills[name] = skill_config

    return skills


def load_treatments_yaml(path: Path | None = None) -> dict[str, TreatmentConfig]:
    """Load treatment configurations from YAML file.

    Args:
        path: Optional path to treatments.yaml

    Returns:
        Dict mapping treatment names to TreatmentConfig objects
    """
    path = path or TREATMENTS_FILE

    if not path.exists():
        raise FileNotFoundError(f"Treatments file not found: {path}")

    with open(path) as f:
        data = yaml.safe_load(f)

    treatments = {}
    for name, cfg in data.items():
        treatments[name] = TreatmentConfig(
            name=name,
            description=cfg.get("description", ""),
            claude_md=cfg.get("claude_md", ""),
            skills=cfg.get("skills", []),
        )

    return treatments


def load_treatment(name: str, path: Path | None = None) -> "Treatment":
    """Load a single treatment by name and build it.

    Args:
        name: Treatment name
        path: Optional path to treatments.yaml

    Returns:
        Built Treatment object ready for use
    """
    from scaffold import Treatment

    configs = load_treatments_yaml(path)

    if name not in configs:
        raise KeyError(f"Treatment not found: {name}. Available: {list(configs.keys())}")

    cfg = configs[name]
    skills = build_treatment_skills(cfg.skills) if cfg.skills else {}

    return Treatment(
        description=cfg.description,
        skills=skills,
        claude_md=cfg.claude_md if cfg.claude_md else None,
        validators=[],  # Validators come from the task, not treatment
    )


def list_treatments(path: Path | None = None) -> list[str]:
    """List available treatment names.

    Args:
        path: Optional path to treatments.yaml

    Returns:
        List of treatment names
    """
    configs = load_treatments_yaml(path)
    return list(configs.keys())

File: scaffold/typescript/treatments.ts
/**
 * Treatment loader for benchmark experiments.
 *
 * Treatments are loaded from YAML configuration and built into Treatment objects
 * with fully-resolved skill configurations.
 *
 * ## Treatment Organization
 *
 * Treatments are organized in the `treatments/` folder by category:
 * - `common/` - Shared treatments (CONTROL, ALL_MAIN_SKILLS, etc.)
 * - `langsmith/` - LS_* treatments for LangSmith tasks
 * - `langchain_concise/` - LCC_* treatments for LangChain tasks
 * - `oss_split/` - OSSS_* treatments for OSS fix tasks (granular skills)
 * - `oss_merged/` - OSSM_* treatments for OSS fix tasks (consolidated skills)
 *
 * All treatments are shared across tasks - there are no task-specific treatments.
 *
 * @example
 * import { loadTreatments, buildTreatmentSkills } from "./treatments.js";
 *
 * const treatments = loadTreatments();
 * const skills = buildTreatmentSkills(treatment.skills);
 */

import { existsSync, readdirSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { parse as parseYaml } from "yaml";
import type { NoiseTask } from "./schema.js";
import { NOISE_TASK_PROMPTS, NOISE_TASK_DELIVERABLES } from "./validation.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

export const TREATMENTS_FOLDER = join(__dirname, "..", "..", "treatments");
export const SKILL_BASE = join(__dirname, "..", "..", "skills", "benchmarks");
export const MAIN_SKILL_BASE = join(__dirname, "..", "..", "skills", "main");
export const NOISE_SKILL_BASE = join(__dirname, "..", "..", "skills", "noise");

// All treatment categories (all shared, no task-specific folders)
const TREATMENT_CATEGORIES = new Set([
  "common",
  "langsmith",
  "langchain_concise",
  "oss_split",
  "oss_merged",
]);

// =============================================================================
// TYPES
// =============================================================================

export interface SkillConfigInput {
  skill?: string;
  name?: string;
  variant?: "py" | "ts" | "all";
  suffix?: boolean;
  noise?: boolean;
  content?: string;
  include_related?: boolean;
  included_sections?: string[];
  extra_sections?: string[];
  section_overrides?: Record<string, string>;
  base?: "benchmarks" | "main";
}

export interface TreatmentConfig {
  name: string;
  description: string;
  claude_md?: string;
  skills?: SkillConfigInput[];
  noise_tasks?: string[];
}

export interface BuiltSkillConfig {
  sections: string[];
  scriptsDir?: string;
  scriptFilter?: string;
}

// =============================================================================
// FUNCTIONS
// =============================================================================

/**
 * Load treatment configurations from a YAML file.
 */
export function loadTreatmentsYaml(path: string): Record<string, TreatmentConfig> {
  if (!existsSync(path)) {
    return {};
  }

  const content = readFileSync(path, "utf8");
  const data = parseYaml(content) as Record<string, unknown>;

  const treatments: Record<string, TreatmentConfig> = {};
  for (const [name, cfg] of Object.entries(data)) {
    // Skip internal keys (anchors) starting with _
    if (name.startsWith("_")) continue;

    const config = cfg as Record<string, unknown>;
    treatments[name] = {
      name,
      description: (config.description as string) || "",
      claude_md: config.claude_md as string | undefined,
      skills: config.skills as SkillConfigInput[] | undefined,
      noise_tasks: config.noise_tasks as string[] | undefined,
    };
  }

  return treatments;
}

/**
 * Load all treatments from the treatments/ folder structure.
 */
export function loadTreatments(): Record<string, TreatmentConfig> {
  if (!existsSync(TREATMENTS_FOLDER)) {
    return {};
  }

  const treatments: Record<string, TreatmentConfig> = {};

  for (const category of readdirSync(TREATMENTS_FOLDER, { withFileTypes: true })) {
    if (!category.isDirectory()) continue;
    if (!TREATMENT_CATEGORIES.has(category.name)) continue;

    const categoryPath = join(TREATMENTS_FOLDER, category.name);
    for (const file of readdirSync(categoryPath)) {
      if (!file.endsWith(".yaml")) continue;

      const yamlPath = join(categoryPath, file);
      const categoryTreatments = loadTreatmentsYaml(yamlPath);
      Object.assign(treatments, categoryTreatments);
    }
  }

  return treatments;
}

/**
 * List available treatment names.
 */
export function listTreatments(): string[] {
  const treatments = loadTreatments();
  return Object.keys(treatments);
}

/**
 * Load treatments available for a task.
 *
 * Note: All treatments are now shared. This function returns all treatments
 * regardless of taskPath (kept for backward compatibility).
 */
export function loadTaskTreatments(
  taskPath: string,
): Record<string, TreatmentConfig> {
  return loadTreatments();
}

/**
 * Get list of treatment names available for a task.
 *
 * Note: All treatments are now shared. This function returns all treatments
 * regardless of taskPath (kept for backward compatibility).
 */
export function getTaskTreatmentNames(taskPath: string): string[] {
  return listTreatments();
}

/**
 * Build NoiseTask objects from task names.
 */
export function buildNoiseTasks(noiseTaskNames: string[]): NoiseTask[] {
  return noiseTaskNames
    .filter((name) => name in NOISE_TASK_PROMPTS)
    .map((name) => ({
      prompt: NOISE_TASK_PROMPTS[name],
      deliverables: [NOISE_TASK_DELIVERABLES[name] || ""],
    }));
}

/**
 * Build skill configuration from YAML skill configs.
 * Note: This is a simplified version - full skill loading requires
 * the skills/parser module which is Python-only.
 */
export function buildTreatmentSkills(
  skillConfigs: SkillConfigInput[] | undefined,
): Record<string, BuiltSkillConfig> {
  if (!skillConfigs) return {};

  const skills: Record<string, BuiltSkillConfig> = {};

  for (const cfg of skillConfigs) {
    const name = cfg.name || cfg.skill?.replace(/_/g, "-") || "unknown";

    // Option 1: Inline content
    if (cfg.content) {
      skills[name] = { sections: [cfg.content] };
      continue;
    }

    // Option 2: Load from skill directory
    if (cfg.skill) {
      const base = cfg.base || "benchmarks";
      const baseDir = base === "main" ? MAIN_SKILL_BASE : SKILL_BASE;
      const skillDir = cfg.noise
        ? join(NOISE_SKILL_BASE, cfg.skill)
        : join(baseDir, cfg.skill);

      // Try to load skill.md content directly
      const skillMdPath = join(skillDir, "skill.md");
      const skillAllPath = join(skillDir, "skill_all.md");
      const skillPyPath = join(skillDir, "skill_py.md");
      const skillTsPath = join(skillDir, "skill_ts.md");
      const skillUpperPath = join(skillDir, "SKILL.md");

      let content = "";
      const variant = cfg.variant || "all";

      if (variant === "py" && existsSync(skillPyPath)) {
        content = readFileSync(skillPyPath, "utf8");
      } else if (variant === "ts" && existsSync(skillTsPath)) {
        content = readFileSync(skillTsPath, "utf8");
      } else if (existsSync(skillAllPath)) {
        content = readFileSync(skillAllPath, "utf8");
      } else if (existsSync(skillUpperPath)) {
        content = readFileSync(skillUpperPath, "utf8");
      } else if (existsSync(skillMdPath)) {
        content = readFileSync(skillMdPath, "utf8");
      }

      if (content) {
        skills[name] = {
          sections: [content],
          scriptsDir: existsSync(join(skillDir, "scripts"))
            ? join(skillDir, "scripts")
            : undefined,
          scriptFilter: variant !== "all" ? variant : undefined,
        };
      }
    }
  }

  return skills;
}
